lib/plugins/wallet/cardano/api.js

```diff
@@ -0,0 +1,161 @@
+          ? [...response.data.data, ...appendData]
+          : response.data.data
+
+        if (
+          method === 'GET' &&
+          response.data.meta.pagination.totalPages >
+            response.data.meta.pagination.page
+        ) {
+          return this.fetch(
+            endpoint,
+            {
+              ...params,
+              page: (params.page || 1) + 1
+            },
+            data,
+            responseData
+          )
+        }
+
+        return responseData
+      })
+  }
+
+  async walletBalance(walletId, index, confirmations = 1) {
+    return this.fetch('transactions', {
+      walletId,
+      accountIndex: index
+    })
+      .then(transactions =>
+        transactions.reduce((acc, item) => {
+          if (item.confirmations < confirmations) {
+            return acc
+          }
+
+          return acc + (item.direction === 'outgoing' ? -1 : 1) * item.amount
+        }, 0)
+      )
+      .then(r => BN(r).round())
+  }
+
+  async addressBalance(address, confirmations) {
+    return this.fetch('transactions', {
+      address: address
+    })
+      .then(transactions =>
+        transactions.reduce((acc, item) => {
+          const inputs = item.inputs
+            .filter(
+              record =>
+                record.address === address &&
+                item.confirmations >= confirmations
+            )
+            .reduce((acc, item) => acc + item.amount, 0)
+
+          const outputs = item.outputs
+            .filter(
+              record =>
+                record.address === address &&
+                item.confirmations >= confirmations
+            )
+            .reduce((acc, item) => acc + item.amount, 0)
+
+          return acc + outputs - inputs
+        }, 0)
+      )
+      .then(balance =>
+        BN(balance)
+          .shift(unitScale)
+          .round()
+      ) // TODO: is that shift really need?
+  }
+
+  async sendCoins(walletId, accountIndex, address, cryptoAtoms) {
+    return this.fetch('transactions', null, {
+      destinations: [
+        {
+          amount: cryptoAtoms.toNumber(),
+          address
+        }
+      ],
+      source: {
+        accountIndex,
+        walletId
+      },
+      groupingPolicy: 'OptimizeForHighThroughput'
+    }).then(data => {
+      const inputs = data.inputs.reduce((acc, item) => acc + item.amount, 0)
+      const outputs = data.outputs.reduce((acc, item) => acc + item.amount, 0)
+
+      return {
+        fee: BN(inputs - outputs)
+          .abs()
+          .shift(unitScale)
+          .round(), // TODO: is that shift really need?
+        txid: data.id
+      }
+    })
+  }
+
+  async newWallet(mnemonic, name = 'Server Wallet', operation = 'create') {
+    return this.fetch('wallets', null, {
+      assuranceLevel: 'normal',
+      backupPhrase: mnemonic,
+      name,
+      operation
+    })
+      .then(data => data.id)
+      .catch(e => {
+        if (e.response.status === 403) {
```

Comment:
could this throw? I mean, `e.response` might not be defined

Reply:
i suppose it's exception from axios, but you are right, i will check this before

Reply:
But if it fails it is fine right? As we would want to throw anyway I guess.